### ISSUE TITLE
leader key in elfeed not recognised

### ISSUE THREAD

**Issue by YiLiu6240** [syl20bnr/spacemacs#7111]
I am trying to add the evilified keys defined in elfeed layer to be invoked by the leader key.

What I did is add the following to `dotspacemacs/user-config`: 

```
  (with-eval-after-load 'elfeed
    (spacemacs/set-leader-keys-for-major-mode 'elfeed-search-mode
      "c"  'elfeed-db-compact
      "gr" 'elfeed-update
      "gR" 'elfeed-search-update--force))
```

However when pressing `,` in elfeed, it says ", is undefined".

Is there a way I can use evil leader key in elfeed?

**Comment by TheBB** [comment]
The leader keys aren't bound in evilified state, reason being that in evilified state the idea is to just bind the keys directly since you don't have to worry so much about shadowing other bindings. So I would just do that instead.

```elisp
  (evil-define-key 'evilified elfeed-search-mode-map "c" 'elfeed-db-compact)
```

and so on.

**Comment by TheBB** [comment]
Looks like I may be mistaken. I don't know why it doesn't work. But anyway just binding the keys directly looks better to me.

**Comment by YiLiu6240** [comment]
I would prefer to leave the evilified state bindings to be as few as possible, except for bindings for navigations, and try to put the bindings for other functionalities in a major mode leader key maps. In this way I can avoid hitting these functionalities by mistake and maintain a clearer functionality structure for a major mode.

What I end up doing is like the following, which mimics a major mode keymap, but it is a bit hacky.
```
(with-eval-after-load 'elfeed
(evilified-state-evilify-map elfeed-search-mode-map
      :mode elfeed-search-mode
      :eval-after-load elfeed-search
      :bindings
      ",c"  'elfeed-db-compact
      ",gr" 'elfeed-update
      ",gR" 'elfeed-search-update--force
      ",gu" 'elfeed-unjam
      "c" nil
      "gr" nil
      "gR" nil
      "gu" nil
      "o" 'elfeed-search-show-entry
      "O" 'elfeed-search-browse-url
      "q"  'quit-window
      "w"  'nil
      "W"  'nil))
```

You said the major mode leader key is not bound in evilified state, but I also had a similar major mode leader key map that works fine for dired, which is in evilified state by default?

**Comment by TheBB** [comment]
> You said the major mode leader key is not bound in evilified state

I also said I was mistaken?

**Comment by YiLiu6240** [comment]
I am just trying to find out what inhibits the invoking of the leader key.

magit-mode also does not recognise the leader key. 
```
(with-eval-after-load 'magit
  (spacemacs/set-leader-keys-for-major-mode 'magit-mode
    "q" 'quit-window))
```

The above also does not work.  In normal state (default) magit treats "," as `evil-repeat-find-char-reverse`, and in emacs state pressing `SPC m` magit says "SPC m is undefined".

Is there a way to know on what condition can/cannot a major mode respect the leader key?

**Comment by TheBB** [comment]
@justbur Do you know anything about this? From what I can tell the bind-map calls are as they should be.

**Comment by justbur** [comment]
I looked at the elfeed one. Check out the `elfeed-search-mode` function. It's defined as an ad hoc major mode and the proper hooks are not being run.

**Comment by TheBB** [comment]
Oh, it's that thing again. Thanks.

@YiLiu6240 Then this is elfeed's problem. You should report there.

**Comment by YiLiu6240** [comment]
I had a look at function definitions of `elfeed-search-mode`, `magit-mode`, and other major modes like `dired-mode` and `python-mode`, but understanding how they work is beyond me right now.

What kind of specific issue is there with elfeed, that causes spacemacs' leader key settings to not work as expected?

**Comment by TheBB** [comment]
Major modes should be defined with `define-derived-mode`. This creates a function with that mode's name, which does what it needs to do to set up the mode. Elfeed instead simply creates the function manually, with the upshot that they've forgotten to do certain things, in this particular case they are not running `change-major-mode-after-body-hook`, which is what bind-map needs for its functionality.

**Comment by atheriel** [comment]
I patched upstream Elfeed (see skeeto/elfeed#214) to call `change-major-mode-after-body-hook` correctly. Don't know if that resolves this issue, but it might help.
